### PR TITLE
ColdBrew | UPMARK-15755 | Capstone Part 10: Repository Pattern

### PR DESCRIPTION
**Pull request by nathanachang**
part10
part10details

### CODE REVIEW

File: app/src/main/java/com/capstone/weatherapp/CityListViewModel.kt [hunk 1]
```diff
@@ -1,26 +1,45 @@
...
 
 private const val CITIES = "4772354,5368361,5128581,4887398,4699066,5809844,5856195,5419384,5391959,4684888,5746545,4990729,4335045"
 private const val APIKEY = "6849ae760f417fb8188f4bb7fd0d92fc"
 private const val UNITS = "imperial"
 
-class CityListViewModel : ViewModel() {
-    private val _cityList = MutableLiveData<List<City>>()
+class CityListViewModel(private val weatherRepo: WeatherRepository) : ViewModel() {
+    private val _cityList = weatherRepo.cityListData
     val cityList: MutableLiveData<List<City>>
```
Comment (katzhm) [inline app/src/main/java/com/capstone/weatherapp/CityListViewModel.kt]: My b for not catching this earlier but this should not be mutable, should just be LiveData since we don't want the fragment or any UI to be able to update this value

Comment (katzhm) [reply app/src/main/java/com/capstone/weatherapp/CityListViewModel.kt]: Actually, you don't even need the private _cityList anymore, you should just need a public cityList that's set to weatherRepo.cityListData. Since you're just setting it to the value of another variable from the repository, don't need the private backing variable anymore

File: app/src/main/java/com/capstone/weatherapp/CityListViewModel.kt [hunk 2]
```diff
@@ -1,26 +1,45 @@
...
+                    _eventNetworkError.value = true
+                }
             }
         }
     }
+}
+
+class CityListViewModelFactory(private val repo: WeatherRepository) : ViewModelProvider.Factory {
+    override fun <T : ViewModel> create(modelClass: Class<T>): T {
+        if (modelClass.isAssignableFrom(CityListViewModelFactory::class.java)) {
```
Comment (katzhm) [inline app/src/main/java/com/capstone/weatherapp/CityListViewModel.kt]: this parameter shouldn't be the `CityListViewModelFactory::class.java` class, it should just be the viewModel (so `CityListViewModel::class.java`

File: app/src/main/java/com/capstone/weatherapp/DetailsViewModel.kt [hunk 3]
```diff
@@ -1,26 +1,45 @@
...
 
 private const val APIKEY = "6849ae760f417fb8188f4bb7fd0d92fc"
 private const val UNITS = "imperial"
 
-class DetailsViewModel : ViewModel() {
-    private val _city = MutableLiveData<SingleCityResponse?>()
-    val city: MutableLiveData<SingleCityResponse?>
+class DetailsViewModel(private val weatherRepo: WeatherRepository) : ViewModel() {
+    private val _city = weatherRepo.singleCityData
+    val city: MutableLiveData<SingleCityResponse>
```
Comment (katzhm) [inline app/src/main/java/com/capstone/weatherapp/DetailsViewModel.kt]: Same thing as the above comment from the other viewModel, don't need the private one anymore

File: app/src/main/java/com/capstone/weatherapp/DetailsViewModel.kt [hunk 4]
```diff
@@ -1,26 +1,45 @@
...
+                    _eventNetworkError.value = true
+                }
             }
         }
     }
+}
+
+class DetailsViewModelFactory(private val repository: WeatherRepository) : ViewModelProvider.Factory {
+    override fun <T : ViewModel> create(modelClass: Class<T>): T {
+        if (modelClass.isAssignableFrom(DetailsViewModel::class.java)) {
```
Comment (katzhm) [inline app/src/main/java/com/capstone/weatherapp/DetailsViewModel.kt:37]: 👍

File: app/src/main/java/com/capstone/weatherapp/WeatherCache.kt [hunk 5]
```diff
@@ -14,25 +14,34 @@ data class CityCache(
...
-                main = Main(temp=temp, feels_like=temp, temp_min=temp_min, temp_max=temp_max, pressure=constants.DOUBLE_ZERO, humidity=humidity),
-                visibility = constants.INT_ZERO,
-                wind = Wind(constants.DOUBLE_ZERO, constants.INT_ZERO),
-                clouds = Clouds(constants.INT_ZERO),
-                dt = constants.INT_ZERO,
-                sys = Sys(country=country, timezone=0, sunrise=0, sunset=0)
-            )
-            cityList.add(city)
-        }
+    companion object {
```
Comment (katzhm) [inline app/src/main/java/com/capstone/weatherapp/WeatherCache.kt:17]: Why did you move this inside a companion object?

File: app/src/main/java/com/capstone/weatherapp/WeatherCacheApplication.kt [hunk 6]
```diff
@@ -4,4 +4,5 @@ import android.app.Application
 
 class WeatherCacheApplication : Application() {
     val db: WeatherDatabase by lazy { WeatherDatabase.getDatabase(this) }
+    val repo: WeatherRepository by lazy { WeatherRepository(db) }
```
Comment (katzhm) [inline app/src/main/java/com/capstone/weatherapp/WeatherCacheApplication.kt:7]: 👍

File: app/src/main/java/com/capstone/weatherapp/WeatherRepository.kt [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+private const val CITIES = "4772354,5368361,5128581,4887398,4699066,5809844,5856195,5419384,5391959,4684888,5746545,4990729,4335045"
+private const val APIKEY = "6849ae760f417fb8188f4bb7fd0d92fc"
+private const val UNITS = "imperial"
+
+class WeatherRepository(private val db: WeatherDatabase) {
+    private val _cityListData = MutableLiveData<List<City>>()
+    val cityListData: MutableLiveData<List<City>>
```
Comment (katzhm) [inline app/src/main/java/com/capstone/weatherapp/WeatherRepository.kt]: this should just be LiveData, not mutable

File: app/src/main/java/com/capstone/weatherapp/WeatherRepository.kt [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        get() = _cityListData
+    private val _singleCityData = MutableLiveData<SingleCityResponse>()
+    val singleCityData: MutableLiveData<SingleCityResponse>
```
Comment (katzhm) [inline app/src/main/java/com/capstone/weatherapp/WeatherRepository.kt]: same as above comment, we don't want this mutable

**Review by katzhm — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/capstone/weatherapp/WeatherCache.kt (hunk 5 above)
Comment (nathanachang) [reply app/src/main/java/com/capstone/weatherapp/WeatherCache.kt:17]: seemed like more of a static class level function to me; we're passing in a list of CityCache objects to be converted, so it's not exactly object specific

Comment (katzhm) [reply app/src/main/java/com/capstone/weatherapp/WeatherCache.kt:17]: Gotcha. If you take it out of the companion object you should just be able to call the function on any instances of the data class. For example:
```
val weatherCache: WeatherCache
weatherCache.convertToWeatherApiResponse()
```
Then you don't need to pass in any parameters to the method and it's a lot cleaner call. You don't need to fix it to this, just wanted to point it out

File: app/src/main/java/com/capstone/weatherapp/WeatherCache.kt [hunk 9]
```diff
@@ -14,25 +14,34 @@ data class CityCache(
...
+                        temp_min=cityCache.temp_min,
+                        temp_max=cityCache.temp_max,
+                        pressure=constants.DOUBLE_ZERO,
+                        humidity=cityCache.humidity
+                    ),
+                    visibility = constants.INT_ZERO,
+                    wind = Wind(constants.DOUBLE_ZERO, constants.INT_ZERO),
+                    clouds = Clouds(constants.INT_ZERO),
+                    dt = constants.INT_ZERO,
+                    sys = Sys(country=cityCache.country, timezone=0, sunrise=0, sunset=0)
```
Comment (katzhm) [inline app/src/main/java/com/capstone/weatherapp/WeatherCache.kt]: let's replace these 0's with the constant

File: app/src/main/java/com/capstone/weatherapp/WeatherCache.kt [hunk 10]
```diff
@@ -57,14 +67,32 @@ data class SingleCityCache(
...
-            weather = emptyList(),
-            base = base,
-            main = Main(temp=temp, feels_like=temp, temp_min=temp_min, temp_max=temp_max, pressure=pressure, humidity=humidity),
+            weather = listOf(Weather(
+                id=constants.INT_ZERO,
+                main=desc,
+                description=desc,
+                icon=icon
+            )),
+            base = "",
```
Comment (katzhm) [inline app/src/main/java/com/capstone/weatherapp/WeatherCache.kt]: let's make this empty string a constant as well

File: app/src/main/java/com/capstone/weatherapp/WeatherRepository.kt [hunk 11]
```diff
@@ -26,19 +26,24 @@ class WeatherRepository(private val db: WeatherDatabase) {
 
     suspend fun refreshCityList() {
         withContext(Dispatchers.IO) {
-            val cityList = WeatherApiClient.retrofitService.getCities(CITIES, APIKEY, UNITS)
-            db.cityDao().insert(cityList.convertToCityCacheList())
+            try {
+                val cityList = WeatherApiClient.retrofitService.getCities(CITIES, APIKEY, UNITS)
+                db.cityDao().insert(cityList.convertToCityCacheList())
+            } catch (e: Exception) {}
+            _cityListData.postValue(getNetworkCityList())
```
Comment (katzhm) [inline app/src/main/java/com/capstone/weatherapp/WeatherRepository.kt:33]: we don't want this inside the catch block because if the try block code is sucessful, this gets skipped over. let's move this back to where it was so that it gets called both if the try block code succeeds or fails

File: app/src/main/java/com/capstone/weatherapp/WeatherRepository.kt [hunk 12]
```diff
@@ -26,19 +26,24 @@ class WeatherRepository(private val db: WeatherDatabase) {
...
 
     suspend fun refreshSingleCity(cityId: String) {
         withContext(Dispatchers.IO) {
-            val singleCity = WeatherApiClient.retrofitService.getCity(cityId, APIKEY, UNITS)
-            db.singleCityDao().insert(singleCity.convertToSingleCityCache())
+            try {
+                val singleCity = WeatherApiClient.retrofitService.getCity(cityId, APIKEY, UNITS)
+                db.singleCityDao().insert(singleCity.convertToSingleCityCache())
+            } catch (e: Exception) {}
+            val singleCityCache = db.singleCityDao().getSingleCityById(cityId)
```
Comment (katzhm) [inline app/src/main/java/com/capstone/weatherapp/WeatherRepository.kt:43]: see above comment - same here
